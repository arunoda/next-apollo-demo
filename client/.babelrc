{
  "presets": ["next/babel"],
  "env": {
    "production": {
      "plugins": [
        ["react-remove-properties", {"properties": ["data-testid"]}]
      ]
    }
  }
}
