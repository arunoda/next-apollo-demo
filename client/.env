NEXT_PUBLIC_GRAPHQL_SERVER=http://localhost:5000/graphql