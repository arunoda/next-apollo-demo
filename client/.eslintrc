{
  "env": {
    "browser": true,
    "commonjs": true,
    "es6": true,
    "node": true
  },
  "extends": [
    "airbnb",
    "airbnb/hooks",
    "eslint:recommended",
    "plugin:react/recommended",
    "plugin:import/errors",
    "plugin:import/warnings",
    "plugin:jsx-a11y/recommended",
    // "plugin:react-hooks/recommended",
    // always put prettier at last
    "prettier"
  ],
  "globals": {
    "Atomics": "readonly",
    "SharedArrayBuffer": "readonly"
  },
  "parserOptions": {
    "ecmaFeatures": {
      "jsx": true // enable linting for jsx files
    },
    "ecmaVersion": 11,
    "sourceType": "module"
  },
  "settings": {
    "react": {
      "version": "detect"
    }
    // "import/resolver": {
    //   "alias": { "map": [["src", "./src"]], "extensions": [".jsx", ".js"] }
    // }
  },
  "plugins": ["react", "react-hooks"],
  "rules": {
    "react/jsx-props-no-spreading": 0,
    "react/forbid-prop-types": 0
  }
}
