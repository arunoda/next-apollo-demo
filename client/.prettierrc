{
  "arrowParens": "avoid",
  "semi": true,
  "trailingComma": "all",
  "singleQuote": true,
  "printWidth": 120
}