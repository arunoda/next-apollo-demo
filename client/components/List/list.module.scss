.cardContainer {
  padding: 8px;
  margin-bottom: 8px;;
  border: 1px solid;
}