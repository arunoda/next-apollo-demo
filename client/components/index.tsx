export { default as Name } from './Name'
export { default as UserTile } from './UserTile'