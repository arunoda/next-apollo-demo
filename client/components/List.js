import React, { useState, useEffect } from "react"
import { gql, useQuery } from '@apollo/client'
import Card from './Card'
import Search from './Search'
import styles from './List.module.css'

const Component = () => {
  let [users, setUsers] = useState([]);

  const query = gql`
    {
      users {
        name
        address
        email
        phone
      }
    }
  `
  let { loading, error, data } = useQuery(query)

  useEffect(() => {
      if (loading === false && data && data.users) {
        setUsers(data.users)
      }
  }, [loading, data])

  if (loading) return 'Loading...'
  if (error) return `Error! ${error.message}`

  return (
    <div className="cards">
      <Search onSearch={setUsers}/>
      <div className="cards__render">
        { users.length > 0
          ? users.map(user => {
          return (
            <div key={user.email}>
              <Card user={user} />
            </div>
          )
        })
        : <div className={styles.none}>No results</div>
      }
      </div>
    </div>
  )
}

export default Component
