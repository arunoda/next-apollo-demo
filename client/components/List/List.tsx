import React from 'react';
import { useEffect, useState } from 'react';
import { useLazyQuery } from "@apollo/client";
import { SearchBar } from '../../components/SearchBar/SearchBar';
import { GET_DETAILS } from '../../apollo/queries/getDetails';

import styles from './list.module.scss';

export const List = () => {
  const [getDetailsQuery, { data, loading, fetchMore }] = useLazyQuery(GET_DETAILS);
  const [detailsArray, setDetailsArray] = useState([]);
  const [inputValue, setInputValue] = useState('');
  const [shouldFetchMore, setShouldFetchMore] = useState(false);

  useEffect(() => {
    setDetailsArray(data?.details || []);
    data?.details.length < 20 ? setShouldFetchMore(false) : setShouldFetchMore(true);
  }, [data])

  useEffect(() => {
    getDetailsQuery({
      variables: {
        offset: 0,
        limit: 20,
        search: inputValue
      }
    })
  }, [inputValue])

  const fetchMoreData = async () => {
    const { data: newData } = await fetchMore({
      variables: {
        limit: 20,
        search: inputValue,
        offset: detailsArray.length
      },
    })
    newData?.details.length < 20 ? setShouldFetchMore(false) : setShouldFetchMore(true);
    setDetailsArray([...detailsArray, ...newData.details])
  }

  return (
    <div>
      <SearchBar setInputValue={setInputValue} />
      {inputValue !== '' && <div className={styles.resultsLabel}>Results for : {inputValue}</div>}

      <div className={styles.listContainer}>
        {loading ? <div>loading</div>
          : detailsArray?.map((detail, index: number) => (
            <div className={styles.cardContainer} key={`${detail?.name}-${index}`} data-testid='cardContainer'>
              <div>{detail?.name}</div>
              <div>{detail?.address}</div>
            </div>
          ))
        }
      </div>
      {shouldFetchMore &&
        <div className={styles.loadButton} data-testid='loadButton' onClick={fetchMoreData}>Load more data</div>
      }
    </div>
  )
}