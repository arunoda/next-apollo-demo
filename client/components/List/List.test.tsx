import React from 'react';
import { act,render, fireEvent } from '@testing-library/react';
import { useLazyQuery } from "@apollo/client";
import { bigReponse } from '../../cypress/mock/mock';
import { List } from './List';

jest.mock('@apollo/client');

const mockUseLazyQuery = useLazyQuery as jest.MockedFunction<any>;
const getDetailsQuery = jest.fn();
const mockFetchMore = jest.fn(() => ({}));

describe('List', () => {

  beforeEach(() => {
    mockUseLazyQuery.mockReturnValue([
      getDetailsQuery,
      {
        data: {
          details: [
            {
              name: 'Name 1',
              address: 'address 1',
              phone: 'phone 1',
              email: 'email 1'
            },
            {
              name: 'Name 2',
              address: 'address 2',
              phone: 'phone 2',
              email: 'email 2'
            },
            {
              name: 'Name 3',
              address: 'address 3',
              phone: 'phone 3',
              email: 'email 3'
            }
          ]
        },
        fetchMore: mockFetchMore
      }
    ])
  });

  it('should match snapshot', () => {
    const { container } = render(<List />);
    expect(container).toMatchSnapshot();
  })

  it('should not display the load more button when there is less than 20 results', () => {
    const { queryByTestId } = render(<List />);
    expect(queryByTestId('loadButton')).toBeFalsy();
  })

  it('should display the load more button when there is more than 20 results', () => {
    mockUseLazyQuery.mockReturnValue([
      getDetailsQuery, { data: bigReponse.data, fetchMore: mockFetchMore }
    ]);
    const { getByTestId } = render(<List />);
    expect(getByTestId('loadButton')).toBeTruthy();
  })

  it('should call fetchMore when clicking on the button', () => {
    mockUseLazyQuery.mockReturnValue([
      getDetailsQuery, { data: bigReponse.data, fetchMore: mockFetchMore }
    ]);
    const { getByTestId } = render(<List />);
    act(() => {
      fireEvent.click(getByTestId('loadButton'))
    });
    expect(mockFetchMore).toHaveBeenCalledWith({
      variables: {
        limit: 20,
        search: '',
        offset: 21
      }
    })
  })
})