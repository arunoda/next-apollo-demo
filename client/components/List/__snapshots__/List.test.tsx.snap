// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List should match snapshot 1`] = `
<div>
  <div>
    <input
      class="input"
      data-testid="searchBar"
      name="searchBar"
      placeholder="Search..."
      type="text"
    />
    <div
      class="listContainer"
    >
      <div
        class="cardContainer"
        data-testid="cardContainer"
      >
        <div>
          Name 1
        </div>
        <div>
          address 1
        </div>
      </div>
      <div
        class="cardContainer"
        data-testid="cardContainer"
      >
        <div>
          Name 2
        </div>
        <div>
          address 2
        </div>
      </div>
      <div
        class="cardContainer"
        data-testid="cardContainer"
      >
        <div>
          Name 3
        </div>
        <div>
          address 3
        </div>
      </div>
    </div>
  </div>
</div>
`;
